### PR TITLE
Option to enable Avro compression

### PR DESCRIPTION
**Pull request by emlyn**
I've added an option to enable compression on the Avro data loaded into Redshift. I haven't benchmarked it but it does seem that it can have a quite significant impact on performance (a small Spark job with the output loaded into Redshift went from ~5 minutes to ~4 minutes when I enabled snappy compression, and roughly half of that run time was on the Spark job itself).
Ideally setting snappy compression by default would be nice, as it should be available everywhere so it's basically a free performance boost. But enabling it involves modifying the Hadoop configuration, which could possibly interact subtly with other parts of the user's job and lead to hard to diagnose issues down the line.
So I've left it disabled by default, but included it in the examples so that it is easier to find.
I also cleaned up a couple of typos in the README examples while I was modifying it.

### CODE REVIEW

File: src/main/scala/com/databricks/spark/redshift/Parameters.scala [hunk 1]
```diff
@@ -194,5 +195,12 @@ private [redshift] object Parameters extends Logging {
 
       s"aws_access_key_id=$accessKeyId;aws_secret_access_key=$secretAccessKey"
     }
+
+    /**
+     * When nonempty/non-null sets the compression codec to use for writing Avro data.
+     *
+     * Defaults to snappy.
```
Comment (jaley) [inline src/main/scala/com/databricks/spark/redshift/Parameters.scala]: Default is actually disabled now, looks like this doc comment needs an update if we're keeping it disabled by default.

Comment (emlyn) [inline src/main/scala/com/databricks/spark/redshift/Parameters.scala]: Thanks, updated.

File: README.md [hunk 2]
```diff
@@ -272,6 +275,17 @@ It may be useful to have some <tt>GRANT</tt> commands or similar run here when l
 table, the changes will be reverted and the backup table restored if post actions fail.</p>
     </td>
  </tr>
+ <tr>
+    <td><tt>avrocompression</tt></td>
+    <td>No</td>
+    <td>No compression (unless set in Hadoop config)</td>
+    <td>
+<p>Sets the compression codec to use on the Avro data to be loaded into Redshift. This overwrites the <tt>avro.output.codec</tt>
```
Comment (jaley) [inline README.md]: Might be nice to enumerate the available options here, to save people researching how it's plumbed through. I believe `"snappy"`, `"bzip2`and `"gzip"` will all work.

Maybe it's worth us supporting an explicit `"none"` option for the sake of clarity? Now that I think about it, the empty string / null option to disable compression is a bit unnecessarily subtle?

Comment (emlyn) [inline README.md]: I'll add the options to the table, I think another valid one is `"deflate"`.
I'm not sure that an empty string is so bad - leaving compression settings alone is not really the same as `none`, unless `none` would explicitly disable compression.

Comment (emlyn) [inline README.md]: According to this: https://avro.apache.org/docs/1.7.7/spec.html#Object+Container+Files only `"deflate"` and `"snappy"` are available (and `"null"` for no compression).

**Comment by emlyn** [comment]
I just noticed this in the output:
```
INFO Configuration.deprecation: mapred.output.compress is deprecated. Instead, use mapreduce.output.fileoutputformat.compress
INFO Configuration.deprecation: mapred.output.compression.type is deprecated. Instead, use mapreduce.output.fileoutputformat.compress.type
```
Should this be setting the hadoop2 properties? Or could it be used with hadoop1 where they're not available?

**Comment by JoshRosen** [comment]
Hey @emlyn, I'm going to be spending a lot more time working on `spark-redshift` this week and was wondering whether you'd be interested in bringing this patch up to date.  Are you blocked awaiting an answer on the Hadoop configuration issue?  AFAIK Spark itself may be setting some of those legacy properties in order to maintain compatibility with Hadoop 1.x.

**Comment by emlyn** [comment]
I've brought this up to date with `master` (sorry, I rebased instead of merging, not quite thinking straight after being away on holiday for a few weeks...).
In order to enable compression, I'm setting the legacy Hadoop properties ([here](https://github.com/emlyn/spark-redshift/blob/compression/src/main/scala/com/databricks/spark/redshift/RedshiftWriter.scala#L185)), but I'm not sure whether I should follow the deprecation warning advice and use the new properties instead, or use the legacy properties for backwards compatibility (or set both?).

**Comment by JoshRosen** [comment]
Don't worry about rebase vs. merge or the cleanliness of the commit log; I'm going to use a merge script to squash everything into a single commit anyways.

I'll take a closer pass through this later in the afternoon. One high-level comment, though: I think that we'll need to add end-to-end tests for this as part of `RedshiftIntegrationSuite` to make sure that it actually works against a real Redshift cluster. For security reasons, the Travis credential injection won't actually work on pull request builds, so what I'll do is pull your changes into a temporary branch in the `databricks` repo, add the test myself, and check whether it passes.

File: README.md [hunk 3]
```diff
@@ -50,21 +50,22 @@ val sqlContext = new SQLContext(sc)
 // Get some data from a Redshift table
 val df: DataFrame = sqlContext.read
     .format("com.databricks.spark.redshift")
-    .option("url", "jdbc:redshift://redshifthost:5439/database?user=username&password=pass")
-    .option("dbtable" -> "my_table")
-    .option("tempdir" -> "s3://path/for/temp/data")
+    .option("url", "jdbc:postgresql://redshifthost:5439/database?user=username&password=pass")
```
Comment (JoshRosen) [inline README.md]: This should be `redshift`. In an earlier commit, I updated the guide to illustrate use of Amazon's official JDBC connector.

File: README.md [hunk 4]
```diff
@@ -50,21 +50,22 @@ val sqlContext = new SQLContext(sc)
...
 
 df.write
-  .format("com.databricks.spark.redshift")
-    .option("url", "jdbc:redshift://redshifthost:5439/database?user=username&password=pass")
-    .option("dbtable" -> "my_table_copy")
-    .option("tempdir" -> "s3://path/for/temp/data")
-  .mode("error")
-  .save()
+    .format("com.databricks.spark.redshift")
+    .option("url", "jdbc:postgresql://redshifthost:5439/database?user=username&password=pass")
```
Comment (JoshRosen) [inline README.md]: Same here: `redshift://`

File: README.md (hunk 2 above)
Comment (JoshRosen) [inline README.md]: We should probably verify that Redshift supports all of these compression modes. I don't see why it would necessarily support fewer of them, but it's possible. I may be able to take care of this as part of integration testing (I might add a new suite which tests all of the compression options).

**Comment by JoshRosen** [comment]
I've opened the https://github.com/databricks/spark-redshift/compare/avro-compression branch to try to write some integration tests for this.

I'm actually not sure whether we want to update the examples to include this; I think that mentioning it in the docs should be sufficient. If we think that's not going to draw enough attention to this feature then I think we should add a separate "performance tips" heading, since other patches will add similar tuning options.

File: README.md [hunk 5]
```diff
@@ -279,6 +282,24 @@ It may be useful to have some <tt>GRANT</tt> commands or similar run here when l
...
+<p>Sets the compression codec to use on the Avro data to be loaded into Redshift. This overwrites the <tt>avro.output.codec</tt>
+key in the Hadoop configuration with the specified value and also sets <tt>mapred.output.compress = true</tt> and
+<tt>mapred.output.compression.type = BLOCK</tt>. If left unset (or set to null or an empty string) it will leave
+the Hadoop configuration unchanged.</p>
+<p>Valid settings are:</p>
+<ul>
+    <li><tt>""</tt> (default): use compression settings from Hadoop config (usually none unless explicitly set).</li>
+    <li><tt>"snappy"</tt>: use snappy compression.</li>
+    <li><tt>"deflate"</tt>: use deflate (zlib) compression (better ratio but more CPU intensive than snappy).</li>
+    <li><tt>"null"</tt>: disable compression.</li>
```
Comment (JoshRosen) [inline README.md]: Is this true? It looks like this just leaves the default configuration options unchanged.

File: src/main/scala/com/databricks/spark/redshift/RedshiftWriter.scala [hunk 6]
```diff
@@ -180,6 +180,12 @@ class RedshiftWriter(jdbcWrapper: JDBCWrapper) extends Logging {
   def saveToRedshift(sqlContext: SQLContext, data: DataFrame, params: MergedParameters) : Unit = {
     val conn = jdbcWrapper.getConnector(params.jdbcDriver, params.jdbcUrl, new Properties()).apply()
 
+    if (params.avrocompression != null && params.avrocompression.nonEmpty) {
+      val conf = sqlContext.sparkContext.hadoopConfiguration
```
Comment (JoshRosen) [inline src/main/scala/com/databricks/spark/redshift/RedshiftWriter.scala]: Instead of mutating the (effectively) global Hadoop configuration, I think that it would be better to clone this configuration and pass our own local copy throughout the Redshift library.  I'm going to make a separate pull request to perform this cleanup.

Comment (JoshRosen) [inline src/main/scala/com/databricks/spark/redshift/RedshiftWriter.scala]: Actually, there's nowhere else that we mutate this configuration, so this change will have to be part of this patch.

File: README.md (hunk 5 above)
Comment (emlyn) [inline README.md]: It would overwrite whatever value was set before with the string `"null"`, which I believe is a dummy codec that does no compression (at least I think I read that somewhere, but can't find it now).

File: src/main/scala/com/databricks/spark/redshift/RedshiftWriter.scala (hunk 6 above)
Comment (emlyn) [inline src/main/scala/com/databricks/spark/redshift/RedshiftWriter.scala]: Yes, it wasn't immediately obvious to me how to copy and pass through a new config, but that would be much nicer than mutating the global config.

**Comment by emlyn** [comment]
If we are able to enable compression without mutating the global config, I think it might be nice to enable some form of compression by default, as a fast codec like snappy should be a net gain in all cases.

**Comment by JoshRosen** [comment]
I'd like to merge this feature in but there are some other bugfixes that are taking higher priority for me right now. If you have time to work on this, would you be interested in writing a new integration test suite for this? I'd be happy to walk you through how to run the integration tests locally (you'll need to create a new S3 bucket and Redshift database for this, though).

**Comment by emlyn** [comment]
I agree it would be nice to get this merged, but I'm not sure that we're going to have too much time over the next few weeks. I might be able to spend a bit of time on it if it's not too much work - what do you envisage by "writing a new integration test suite"? Is it just testing that importing into Redshift also works when compression is enabled? We already have a dev Redshift cluster that I presume we could use for this, so hopefully setting it up shouldn't be too hard.

**Comment by JoshRosen** [comment]
What I meant was writing some end-to-end tests which exercise this feature against a real Redshift cluster, similar to what we have in `RedshiftIntegrationSuite` (although I'd add a new suite for compression-related tests, similar to what I did with `STSIntegrationSuite`. Given how many bugs I've found in even fairly simple features, I'm fairly hesitant to merge new features which we don't have tests for, especially if it's not _that_ hard to add them (I'd say an hour, tops, for this feature).

**Comment by JoshRosen** [comment]
I'd also be interested in seeing a benchmarking tool / suite / notebook for this, esp. since it would come in handy for #73.

**Comment by emlyn** [comment]
Is this the sort of thing you mean? I can try to run it locally if you give me some pointers on how to run the integration tests.

**Comment by JoshRosen** [comment]
Yep, that's along the lines of what I have in mind. I'm out for the rest of today but I'll try to take a closer look at this tomorrow or Friday.

**Comment by JoshRosen** [comment]
It looks like Snappy compression is enabled by default in the newer version of `spark-avro` that this library uses: https://github.com/databricks/spark-avro/blame/01ae707bf3dbe3e74cfc8b04e892eb34defd28fe/src/main/scala/com/databricks/spark/avro/AvroRelation.scala#L89

Given that it looks like we've been using compression this whole time starting with 0.5.0+, is this PR still relevant?

**Comment by emlyn** [comment]
In that case, I agree that this is probably no longer needed.

**Comment by JoshRosen** [comment]
Great! I'm going to close this as "wontfix" then.

**Comment by jomach** [comment]
@JoshRosen Any updates on this one ? Thanks
